internal/icinga2/client.go

```diff
@@ -213,30 +274,77 @@ func (client *Client) buildDowntimeEvent(ctx context.Context, d Downtime, startE
 		return nil, err
 	}
 
+	var reason string
 	if startEvent {
 		ev.Type = event.TypeDowntimeStart
+		ev.SetMute(true, "Checkable is in downtime")
+		ev.Message = d.Comment
 	} else if !d.WasCancelled() {
 		ev.Type = event.TypeDowntimeEnd
+		reason = "Downtime expired"
 	} else {
 		ev.Type = event.TypeDowntimeRemoved
+		if d.ConfigOwner != "" {
+			reason = fmt.Sprintf("Downtime was cancelled by config owner (%s)", d.ConfigOwner)
+		} else {
+			reason = "Downtime was cancelled by user"
+		}
 	}
 
 	ev.Username = d.Author
-	ev.Message = d.Comment
+	if ev.Type != event.TypeDowntimeStart {
+		queryResult, err := client.fetchCheckableIfUnmuted(ctx, d.Host, d.Service)
+		if err != nil {
+			return nil, err
+		}
+
+		if queryResult == nil {
+			// If queryResult is nil, then the Checkable is still muted due to other reasons, so no need to unmute it.
+			ev.Message = reason
+		} else {
+			// When a downtime is cancelled/expired and there's no other active downtime/ack, we're going to send some
+			// notifications if there's still an active incident. Therefore, we need the most recent CheckResult of
+			// that Checkable to use it for the notifications.
+			ev.Message = queryResult.Attrs.LastCheckResult.Output
+			ev.SetMute(false, reason)
+		}
+	}
 
 	return ev, nil
 }
 
 // buildFlappingEvent from the given fields.
-func (client *Client) buildFlappingEvent(ctx context.Context, host, service string, isFlapping bool) (*event.Event, error) {
-	ev, err := client.buildCommonEvent(ctx, host, service)
+func (client *Client) buildFlappingEvent(ctx context.Context, flapping *Flapping) (*event.Event, error) {
+	ev, err := client.buildCommonEvent(ctx, flapping.Host, flapping.Service)
 	if err != nil {
 		return nil, err
 	}
 
-	ev.Type = event.TypeFlappingStart
-	if !isFlapping {
+	if flapping.IsFlapping {
+		ev.Type = event.TypeFlappingStart
+		ev.SetMute(true, fmt.Sprintf(
+			"Checkable started flapping (Current flapping value %d%% > high threshold %d%%)",
+			flapping.CurrentFlapping, flapping.ThresholdHigh,
+		))
+	} else {
 		ev.Type = event.TypeFlappingEnd
+		queryResult, err := client.fetchCheckableIfUnmuted(ctx, flapping.Host, flapping.Service)
+		if err != nil {
+			return nil, err
+		}
+
+		reason := fmt.Sprintf(
+			"Checkable stopped flapping (Current flapping value %d%% < low threshold %d%%)",
+			flapping.CurrentFlapping, flapping.ThresholdLow,
+		)
+
+		if queryResult == nil {
+			// If queryResult is nil, then the Checkable is still muted due to other reasons, so no need to unmute it.
+			ev.Message = reason
+		} else {
+			ev.Message = queryResult.Attrs.LastCheckResult.Output
+			ev.SetMute(false, reason)
+		}
```

Comment:
> because I missed that you're using the response for more than checking if it's still muted.

That's exactly what I'm doing here too, and to be honest, the previous version was perfect enough for me, I just didn't want to waste time discussing why your suggested solution wasn't better than it was in anyway. We shouldn't over engineer things too much just to *deduplicate* a simple if statement, which by the way wasn't duplicated code at all. These new functions are just useless and overcomplicate things unnecessarily.